audio/audio_hw.c

```diff
@@ -1874,10 +1875,45 @@ static ssize_t do_in_pcm_read(struct audio_stream_in *stream, void *buffer,
         goto exit;
     }
 
-    if (in->resampler.resampler != NULL) {
-        ret = read_resampled_frames(in, buffer, frames_rq);
+    if (in->common.channel_count != in->hw_channel_count) {
+
+        ALOGV("do_in_pcm_read(): Read %d channel stream with adjustment (HW only supports %d channel reading)",
+              in->common.channel_count, in->hw_channel_count);
+
+        size_t read_buf_size = bytes * in->hw_channel_count / in->common.channel_count;
+        void *read_buffer = calloc(1, read_buf_size);
+        if (!read_buffer) {
+            ret = -1;
+            goto exit;
+        }
+
+        if (in->resampler.resampler != NULL) {
+            ret = read_resampled_frames(in, read_buffer, frames_rq);
+        } else {
+            ret = pcm_read(in->pcm, read_buffer, read_buf_size);
+        }
+
```

Comment:
The only difference between lines 1890-1894 and the normal case are the destination buffer and the pcm read size. I would prefer that this remains as one code instance at the top of the function, selecting the correct destination buffer depending on whether a pre-processing buffer was allocated when the stream was opened. The channel adjustment is then done as a post-processing step on the data that was read. So:

if (in->pre_read_buf) {
    buf = in->pre_read_buf;
    buf_size = in->pre_read_buf_size;
} else {
   buf = buffer;
   buf_size = bytes;
}

if (in->resampler.resampler) {
...
} else {
...
}

if (in->common.channel_count != in->hw_channel_count) {
   ...
   adjust_channels(...);
   ...
}

Reply:
@rfvirgil , thanks for your comments, I will think about them.

Reply:
@rfvirgil , I have updated the patch. Please could you look?